### PR TITLE
Triangles

### PR DESCRIPTION
**Pull request by weichweich**
I added an iterator which can draw triangles. It uses the line iterator, to draw the edges. To fill the triangle I start to draw the edges at the top left corner and fill the space between the points on the edge.

Because the current way of drawing lines not always starts drawing at the start point, I had to reimplement the line drawing algorithm.

There are some issues with my triangle implementation. First issue is that triangles have no stroke width. The second issue is that the triangle iterator returns some points multiple times. The corner point e.g. are return at least twice because they are part of two edges which are drawn separately.

### CODE REVIEW

File: embedded-graphics/src/primitives/line.rs [hunk 1]
```diff
@@ -90,121 +90,84 @@ where
...
-            slow,
-            end,
+            style: self.style,
+            
+            start: self.start,
+            end: self.end,
+            d,
+            s,
+            err: d[0] + d[1],
+            e2: 0,
```
Comment (jamwaffles) [inline embedded-graphics/src/primitives/line.rs]: I think these single letters are from Bresenham's algorithm, no? Can you give `d`, `s` and `e2` more descriptive names please?

File: embedded-graphics/src/primitives/line.rs [hunk 2]
```diff
@@ -90,121 +90,84 @@ where
...
-            increment,
-            error,
-
-            quick,
-            slow,
-            end,
+            style: self.style,
+            
+            start: self.start,
+            end: self.end,
```
Comment (jamwaffles) [inline embedded-graphics/src/primitives/line.rs]: These three lines can be replaced with `..self` I think.

File: embedded-graphics/src/primitives/line.rs [hunk 3]
```diff
@@ -90,121 +90,84 @@ where
...
 // [Bresenham's line algorithm](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm)
-impl<'a, C: PixelColor> Iterator for LineIterator<'a, C> {
+impl<C: PixelColor> Iterator for LineIterator<C> {
     type Item = Pixel<C>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        if self.quick > self.end {
-            return None;
-        }
+        self.style.stroke_color?;
```
Comment (jamwaffles) [inline embedded-graphics/src/primitives/line.rs]: Can you add a quick comment explaining that this bails early if there's no stroke? It took me a moment to figure out so might confuse others too. Quite an elegant solution though, nice!

File: simulator/examples/fill.rs [hunk 4]
```diff
@@ -6,14 +6,14 @@ use std::time::Duration;
...
-use embedded_graphics::primitives::{Circle, Rect};
+use embedded_graphics::primitives::{Circle, Rect, Triangle};
 
 use simulator::DisplayBuilder;
 
 static CIRCLE_SIZE: i32 = 32;
 
 fn main() {
-    let mut display = DisplayBuilder::new().size(256, 128).scale(2).build();
+    let mut display = DisplayBuilder::new().size(304, 128).scale(2).build();
```
Comment (jamwaffles) [inline simulator/examples/fill.rs]: Why did this need to change?

File: simulator/src/lib.rs [hunk 5]
```diff
@@ -107,7 +107,7 @@ pub struct DisplayBuilder {
 impl DisplayBuilder {
     pub fn new() -> Self {
         Self {
-            width: 256,
+            width: 320,
```
Comment (jamwaffles) [inline simulator/src/lib.rs]: Why does this need to change?

File: embedded-graphics/src/primitives/line.rs (hunk 2 above)
Comment (weichweich) [reply embedded-graphics/src/primitives/line.rs]: I tried that, but this results in an error. 

```
expected struct `primitives::line::LineIterator`, found struct `primitives::line::Line`
```

File: simulator/examples/fill.rs (hunk 4 above)
Comment (weichweich) [reply simulator/examples/fill.rs]: I added some triangles so I had to increase the window size in order for it to fit all primitives.

File: simulator/src/lib.rs (hunk 5 above)
Comment (weichweich) [reply simulator/src/lib.rs]: I added triangles to the stroke example and wanted to make them fit into the window. I didn't realize that this should be done in the example itself. Should be fixed with my last commit.

**Comment by weichweich** [comment]
> Thanks for the PR! I've left a preliminary review, and I'll test the simulator examples when I'm next at a Linux machine. Can you add a `Triangle` bullet to the `Primitives` list on the [docs homepage](https://docs.rs/embedded-graphics/0.4.7/embedded_graphics/) please?

Done

> If a line has zero length, surely it shouldn't be drawn at all? What's the current behaviour in Embedded Graphics for this? Is there an established/expected behaviour from the wider graphics community?

I didn't put much thought in that. I think the behaviour should not be changed. So zero length lines should no longer be drawn.

> Don't worry about stroke width - that's something that needs to be fixed for the line primitive as well!

Lines with a stroke width are rectangles. If the triangle solution is stable enough, it could be reused for lines and polygons.

File: embedded-graphics/src/primitives/line.rs (hunk 1 above)
Comment (weichweich) [reply embedded-graphics/src/primitives/line.rs]: should be fixed now

File: embedded-graphics/src/primitives/line.rs (hunk 2 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/line.rs]: Oops, you're right

File: simulator/examples/fill.rs (hunk 4 above)
Comment (jamwaffles) [reply simulator/examples/fill.rs]: Ah I see, no problem

**Comment by jamwaffles** [comment]
Released in [0.4.8](https://crates.io/crates/embedded-graphics/0.4.8).

**Pull request merged**